modules/library/render/src/main/java/org/geotools/renderer/lite/MarkFeatureIterator.java

```diff
@@ -0,0 +1,244 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ * 
+ *    (C) 2015, Open Source Geospatial Foundation (OSGeo)
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.renderer.lite;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.logging.Logger;
+
+import org.geotools.data.sort.SimpleFeatureIO;
+import org.geotools.feature.FeatureCollection;
+import org.geotools.feature.FeatureIterator;
+import org.geotools.util.DefaultProgressListener;
+import org.geotools.util.logging.Logging;
+import org.opengis.feature.Feature;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+import org.opengis.util.ProgressListener;
+
+/**
+ * A FeatureIterator that can have a position marked, and can be reset to it
+ * 
+ * @author Andrea Aime - GeoSolutions
+ */
+abstract class MarkFeatureIterator implements FeatureIterator<Feature> {
+
+    static final Logger LOGGER = Logging.getLogger(MarkFeatureIterator.class);
+
+    /**
+     * Builds a new {@link MarkFeatureIterator} making sure no too many features are kept in memory.
+     * The listener won't receive any notification, but will be used to check if the data loading
+     * should be stopped using {@link ProgressListener#isCanceled()}
+     * 
+     * @param fc
+     * @param maxFeaturesInMemory
+     * @return
+     * @throws IOException
+     */
+    public static MarkFeatureIterator getMarkFeatureIterator(FeatureCollection fc,
+            int maxFeaturesInMemory, ProgressListener listener)
+            throws IOException {
```

Comment:
Factory method (switching between private implementations). Often I name factory methods with create (by convention), in this case since it is a static method:

```
MarkFactoryIterator iterator = MarkFactoryIterator.create( features, 10000, null )
```